app/Models/Quiz.php

```diff
@@ -34,21 +36,32 @@ class Quiz extends Model
 
     public function users(): BelongsToMany
     {
-        return $this->belongsToMany(User::class);
+        return $this->belongsToMany(User::class)
+            ->withPivot('score')
+            ->withTimestamps();
     }
 
-    public function questions(): HasMany
+    public function questions(): BelongsToMany
     {
-        return $this->hasMany(Question::class);
+        return $this->belongsToMany(Question::class);
     }
 
-    public function module(): BelongsTo
+    public function course(): BelongsTo
     {
-        return $this->belongsTo(Module::class);
+        return $this->belongsTo(Course::class);
     }
 
-    public function course(): BelongsTo
+    public function getQuestionsInOrder()
     {
-        return $this->belongsTo(Course::class);
+        if ($this->order) {
+            $orderedQuestions = collect([]);
+            foreach ($this->order as $id) {
+                $orderedQuestions->push(Question::find($id));
+            }
+
+            return $orderedQuestions->merge($this->questions)->unique('id')->values();
+        }
+
+        return $this->questions;
     }
```

Comment:
I wonder if you could use [`sort`](https://laravel.com/docs/9.x/collections#method-sort) or [`sortBy`](https://laravel.com/docs/9.x/collections#method-sortby) with a function passed in that returns the index in the order array? Whichever approach is more performant is fine.

Reply:
Hey @jobara when I think about your question again, the lines 56 to 60 are to generate a Question collection with given question ids from $this->order; the logic is not to sort the collection, but generated collection is sorted. The line 62 is to remove the duplication on the collection.

Reply:
Is it possible to use sort or sortBy to generate a collection with given array of ids?

Reply:
I think you'd end up doing something like:

```php
$this->questions->sort(function ($a, $b) {
    if ($a->id === $b->id) { 
        return 0;
    }

    $aIndex = array_search($a->id, $this->order);
    $bIndex = array_search($b->id, $this->order);

    $aIndex = $aIndex === false ? count($this->order) : $aIndex;
    $bIndex = $bIndex === false ? count($this->order) : $bIndex;

    if ($a === false && $b === false) {
        return 0;
    }

    if ($a === false) {
        return 1;
    }

    if ($b === false) {
        return -1;
    }

    return $a < $b ? -1 : 1;
})
```

That looks a bit complicated, but maybe you could optimize it a bit. Not sure.

Reply:
Thanks @jobara, the sort function seems to work for the test case I wrote for the function, but not on the UI; I am not sure what causes the difference what's appear on the UI, but the questions are not ordered according to the order property, but they are ordered correctly by id.